### PR TITLE
feat: DHIS2-9276 #2.1 - api utils for epics

### PR DESCRIPTION
**Pull request by JoakimSM**
[DHIS2-9276](https://jira.dhis2.org/browse/DHIS2-9276)

Main purpose here is to make methods/data from the app-runtime dataEngine available to the epics. You will see it in use in the next PR. Methods/data derived from the dataEngine will be available to the epics through a third parameter (added using the dependencies argument in `getStore.js`). I'm aware that we could potentially have made the dataEngine available through a global, but wanted to test this approach instead for a couple of reasons 1) The dataEngine is already available to the React components through the hook `useDataEngine` 2) Potential Jest tests for the epics should be easier to write when the dataEngine methods are made available through a param.

Sorry about all the changes in the `metaDataStoreLoaders` folder, could have been its own PR, but it was tied to changes I did so opted to do it here. The change here is that I'm passing in a function that queries for a single resource only, since querying for multiple resources are never used.

### CODE REVIEW

File: src/core_modules/capture-core/flow/app-runtime_v2.x.x.js [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    declare type FetchErrorTypeName = 'network' | 'unknown' | 'access' | 'aborted';
+    declare type FetchErrorDetails = { message?: string };
+
+    declare interface FetchErrorPayload {
+        type: FetchErrorTypeName,
+        details?: FetchErrorDetails,
+        message: string,
+    }
+
+    declare class FetchError extends Error implements FetchErrorPayload {
```
Comment (JoakimSM) [inline src/core_modules/capture-core/flow/app-runtime_v2.x.x.js:43]: LGTM made a complaint here, but I believe the code should be ok.

File: src/store/getStore.js [hunk 2]
```diff
@@ -36,14 +35,14 @@ export function getStore(
...
-        effect: getEffectReconciler(onApiMutate),
+        effect: getEffectReconciler(apiUtils.mutate),
         persistCallback: onRehydrated,
         queue: queueConfig,
         persistOptions: getPersistOptions(),
     });
 
     const epicMiddleware = createEpicMiddleware({
-        dependencies: {},
+        dependencies: apiUtils,
```
Comment (JoakimSM) [inline src/store/getStore.js:45]: `apiUtils` made available to the epics through the third parameter.

File: src/components/AppLoader/AppLoader.component.js [hunk 3]
```diff
@@ -15,30 +16,39 @@ type Props = {
...
+        mutate: dataEngine.mutate.bind(dataEngine),
+        absoluteApiPath: buildUrl(dataEngine.link.baseUrl, dataEngine.link.apiPath),
+    }), [dataEngine]);
+};
+
 const AppLoader = (props: Props) => {
     const { onRunApp, onCacheExpired } = props;
     const [loadError, setLoadError] = React.useState(null);
-    const dataEngine = useDataEngine();
+    const apiUtils = useApiUtils();
```
Comment (paschalidi) [inline src/components/AppLoader/AppLoader.component.js:31]: very nitty 
```suggestion
    const { querySingleResource, querySingleResource, apiUtils }  = useApiUtils();
```

File: src/core_modules/capture-core/metaDataStoreLoaders/IOUtils/IOUtils.types.js [hunk 4]
```diff
@@ -1,42 +1,37 @@
...
     iterationSize?: number,
-};
+|};
 
-export type QueryRecursivelyOptions = {
+export type QueryRecursivelyOptions = {|
     convertQueryResponse?: ConvertQueryResponseFn,
     pageSize?: number,
-};
+|};
```
Comment (paschalidi) [inline src/core_modules/capture-core/metaDataStoreLoaders/IOUtils/IOUtils.types.js:37]: ⭐ super

**Review by paschalidi — APPROVED** [review]
Hey thanks for explaining what is going on here. 

Keen to see how we can use the `apiUtils` in the epics.

Great work 👍

File: src/components/AppLoader/AppLoader.component.js (hunk 3 above)
Comment (JoakimSM) [reply src/components/AppLoader/AppLoader.component.js:31]: Makes sense 👍 Changing.

**Pull request merged**